### DISCUSSION TITLE
Loading .ttf font

### DISCUSSION THREAD

**Discussion by zolotaya-ruda** [hecrj/iced#1988] [Q&A] [answered]
How can a load my own font? I'm trying:

```rust
let MY_FONT= Font::External {
                name: "noto-sans-regular",
                bytes: include_bytes!(r"path/to/font.ttf"),
            };
``` 
But as far as I understand from the documentation, the `External `structure is not used now

**Comment by genusistimelord** [comment, marked as answer]
the way we load fonts is very different now which does make it much harder to work with, but here is a working example of loading a font.
https://github.com/iced-rs/iced_aw/blob/main/examples/color_picker/src/main.rs#L48

I just wished we could tell the system what we wanted the font to be called rather than what it loads as.

**Reply by zolotaya-ruda** [reply]
thank you very much!

**Reply by lufte** [reply]
How does it work if I want to load multiple fonts though? In the examples I'm seeing, people define a `Font` with `Font::with_name` and then load a font with `font::load`, but there's no connection between those two things. For example:

```rust
const ICON_FONT_BYTES: &[u8] = include_bytes!("../fonts/icons.ttf");
const ICON_FONT: Font = Font::with_name("icons");
...
font::load(ICON_FONT_BYTES).map(Message::FontLoaded)
```
I first imagined it could be the name, but in same examples not even that matches.

**Reply by fmonniot** [reply]
In my experience the name of the file does not matter. The only thing that matter is what name (and other parameters, like weight & co) is defined in the ttf file itself. You can issue multiple font load commands (batching them if needed) to load multiple fonts.

That's [what I ended up doing](https://github.com/fmonniot/pathfinder-wotr-editor/commit/c86fb9a5d2b77b63f284026de3c269fb798dc9ef#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR106-R116) when upgrading to 0.10.

**Reply by lufte** [reply]
I see, I just used `fc-scan` to inspect the contents of the font files themselves and yes, the "fullname" matches what is passed to `with_name`. I had no idea how that connection was being made so I couldn't imagine how loading multiple fonts would work, as in what font would end up in what variable.

Thanks, your example makes things much clearer.

**Reply by genusistimelord** [reply]
in most cases once the font is loaded you don't even have to choose it if it contains special characters that no other fonts have as it should choose this. But for Unicode and such that already exists you do need to select the font manually. Also the Font name is the Family name generally. this has confused a lot of people. Also you can load multiple fonts that way and manually set them per each thing.  you just do a bunch of font loads one after the other and they each will run and load a font in. the only message that is important it the one that lets the system know they are all loaded.

**Reply by lufte** [reply]
> Also the Font name is the Family name generally

Huh, this does sound confusing. I assume this is not a whim of Iced and it's the way fonts are usually managed, but again I have no idea :)

Also, this being an async command now, I would imagine we can't actually use the fonts until they are loaded, which differs from the previous versions where fonts were loaded at... build time? They were `const`s after all. And therefore they were immediately available.

**Reply by genusistimelord** [reply]
correct that is why the examples are setup with a final command that says it is loaded. until then it should just display a loading page if there is a lot of preloaded parts. This could be made Easier by Iced. simply by making a fonts to load array in the Settings and loading them in a sync way then upon sandbox or application creation. One of the things I would like to have is the ability to set names per loaded font and not use whats inside the file itself.
